### DISCUSSION TITLE
Ibis-Substrait TPC Queries

### DISCUSSION THREAD

**Discussion by gforsyth** [duckdb/duckdb#3282] [Show and tell]
Hi all!

With the recent addition of `from_substrait` and `get_substrait` to DuckDB and the Python client library, we've been experimenting with sending over TPC-H queries from Ibis, serialized via Ibis-substrait, to DuckDB for execution.

Currently, none of the TPC-H queries complete, some because Ibis fails to compile, others because of internal errors in DuckDB, but I think a lot of these can be fixed up without a huge lift (:crossed_fingers: )

There is a table with a brief write-up on the nature of the failures available here: https://github.com/ibis-project/tpc-queries/pull/3

And of the queries that Ibis-Substrait _can_ currently serialize, here is a gist with the serialized bytes that should be copy/pasteable into `duckdb`

https://gist.github.com/gforsyth/734eade9030b256e5641c43dd367c45f

cc @hannes @pdet @cpcloud

**Comment by pdet** [comment]
Hey @gforsyth, thanks for all the effort there.

I've started checking this issue, and is mostly missing types (e.g., tinyints) and function names that currently don't match (e.g., equal != equals), so I have hopes we can get this working soon-ish :-)

I think the function names can get more problematic since systems can (and will) have different names for the same functions. One possibility is to have a soft definition (i.e., everyone follows what Postgres has), or we do an Enum of function names in substrait? Or are there any other ideas?

**Reply by gforsyth** [reply]
Hey @pdet -- glad to hear that it's mostly smallish stuff!  Regarding the function names, I think that an Enum / lookup table of what substrait defines is probably the way to go and then consumers and producers are responsible for mapping their own names onto what substrait calls a given function.
At this point, it's probably also relatively easy to change the names that substrait is using (or at least, not nearly as hard as it will get in the future once adoption picks up).

@cpcloud may have more detailed thoughts on this

**Reply by cpcloud** [reply]
@pdet @gforsyth Let's try and standardize on substrait function names. If a function doesn't exist, we can add it to the spec where possible.

**Comment by pdet** [comment]
Hey @gforsyth 
Already got some of the queries working (yay :)), one minor thing I'm noticing, is that it seems that ibis is producing the queries with the columns in a random-ish order? Or at least not following the original tpc-h query?

e.g.,
For Q01 I get the following values when iterating the aliases in the root node
[0] = {std::string} "l_returnflag"
[1] = {std::string} "l_linestatus"
[2] = {std::string} "avg_disc"
[3] = {std::string} "avg_price"
[4] = {std::string} "avg_qty"
[5] = {std::string} "count_order"
[6] = {std::string} "sum_base_price"
[7] = {std::string} "sum_charge"
[8] = {std::string} "sum_disc_price"
[9] = {std::string} "sum_qty"

while the query itself has the following order:
```sql 
select
	l_returnflag,
	l_linestatus,
	sum(l_quantity) as sum_qty,
	sum(l_extendedprice) as sum_base_price,
	sum(l_extendedprice * (1 - l_discount)) as sum_disc_price,
	sum(l_extendedprice * (1 - l_discount) * (1 + l_tax)) as sum_charge,
	avg(l_quantity) as avg_qty,
	avg(l_extendedprice) as avg_price,
	avg(l_discount) as avg_disc,
	count(*) as count_order
```

**Reply by gforsyth** [reply]
Hey @pdet -- great news re: working queries!

The column order isn't random, but you are right that it doesn't follow the TPC-H query.  There may some other sorting rules, but I think the order for Q01 is because we do a group-by with `l_returnflag` and `l_linestatus`, given in that order, so that order shows up in the column listing, followed by the rest of the columns in alphabetical order.

If we do a reprojection after the aggregation we can select the columns in a specific order.  I'll look into whether this is incidental behavior or a design choice.

**Reply by gforsyth** [reply]
Turns out the alphabetical sorting dates back to when Python dictionary insert ordering wasn't preserved -- we're removing that now so that column ordering can be preserved.

**Comment by pdet** [comment]
Hey @gforsyth,

A bit more development on these queries.

We currently have Q 03,06 and 10 fully functioning (yay).

Q01 seems to present a small issue with the filter the comparison is a < instead of <= 

Q09 Has a cast as a ```ScalarFunction ``` instead of a ```Expression Cast```, we also initially started developing the extension before this definition, in our case, the scalar function would need 2 args (one to define the return type and other the column). 
From ibis, I only get the column here, so I'm not so sure how to transform it. I guess we should both default to the ```Expression Cast``` here anyway.

All queries with subqueries: Have ```Subquery Expressions``` as ```ScalarFunction```, which again I'm not sure how to deserialize them. DuckDB doesn't really export subqueries since they are all flattened, but I guess Ibis should probably serialize subqueries as ```Subquery Expressions```. 

Q19/20: Require a ```value_list``` function, what exactly should this be doing?

**Reply by gforsyth** [reply]
Hey @pdet !  We talked through this in person (woo!) but just writing out some answers here for the record (and so we can refer back).

The `<` instead of `<=` was just a typo in our expression (hah) -- I'll send over the updated proto for that.

We're both going to look at `Expression Cast` for Q09

Ibis needs work before we can detect subqueries during expression creation, so that's on our end

For Q19 and Q20, we shouldn't be passing through `value_list` as that's an Ibis internal, that will be fixed in our substrait compiler.

**Reply by cpcloud** [reply]
For Q19 and Q20, we can specialize our `ops.Contains` to https://github.com/substrait-io/substrait/blob/main/proto/substrait/algebra.proto#L497-L500.

**Reply by gforsyth** [reply]
Ok, I updated the gist https://gist.github.com/gforsyth/734eade9030b256e5641c43dd367c45f with the fixed Q01

**Reply by pdet** [reply]
One other thing is the function names,
I've changed all the duckdb ones to follow the most recent substrait specification.

e.g., + add
>= gte ...
